clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -307,6 +308,7 @@ public static byte requiredMagicForVersion(short produceRequestVersion) {
                 return RecordBatch.MAGIC_VALUE_V1;
 
             case 3:
+            case 4:
```

Comment:
Thanks. Discussed this a bit with Jason. It seems that we probably need to extend the current desired version logic to support the possibility of having multiple versions on the same magic (perhaps having a desired version range). So, we can do this in a followup jira. If you ping Jason on that jira, he can help you with some suggestions.